tower-http/src/metrics/in_flight_requests.rs

```diff
@@ -0,0 +1,283 @@
+        self.count.fetch_add(1, Ordering::Relaxed);
+        IncrementGuard {
+            count: self.count.clone(),
+        }
+    }
+
+    /// Run a future every `interval` which receives the current number of in-flight requests.
+    ///
+    /// This can be used to send the current count to your metrics system.
+    ///
+    /// This function will loop forever so normally it is called with [`tokio::spawn`]:
+    ///
+    /// ```rust,no_run
+    /// use tower_http::metrics::in_flight_requests::InFlightRequestsCounter;
+    /// use std::time::Duration;
+    ///
+    /// let counter = InFlightRequestsCounter::new();
+    ///
+    /// tokio::spawn(
+    ///     counter.run_emitter(Duration::from_secs(10), |count: usize| async move {
+    ///         // Send `count` to metrics system.
+    ///     }),
+    /// );
+    /// ```
+    pub async fn run_emitter<F, Fut>(mut self, interval: Duration, mut emit: F)
+    where
+        F: FnMut(usize) -> Fut + Send + 'static,
+        Fut: Future<Output = ()> + Send,
+    {
+        let mut interval = tokio::time::interval(interval);
+
+        loop {
+            // if all producers have gone away we don't need to emit anymore
+            match Arc::try_unwrap(self.count) {
+                Ok(_) => return,
+                Err(shared_count) => {
+                    self = Self {
+                        count: shared_count,
+                    }
+                }
+            }
+
+            interval.tick().await;
+            emit(self.get()).await;
+        }
+    }
+}
+
+struct IncrementGuard {
+    count: Arc<AtomicUsize>,
+}
+
+impl Drop for IncrementGuard {
+    fn drop(&mut self) {
+        self.count.fetch_sub(1, Ordering::Relaxed);
+    }
+}
+
+impl<S, R, ResBody> Service<R> for InFlightRequests<S>
+where
+    S: Service<R, Response = Response<ResBody>>,
+{
+    type Response = Response<ResponseBody<ResBody>>;
+    type Error = S::Error;
+    type Future = ResponseFuture<S::Future>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: R) -> Self::Future {
+        let guard = self.counter.increment();
+        ResponseFuture {
+            inner: self.inner.call(req),
+            guard: Some(guard),
+        }
+    }
+}
+
+/// Response future for [`InFlightRequests`].
+#[pin_project]
+pub struct ResponseFuture<F> {
+    #[pin]
+    inner: F,
+    guard: Option<IncrementGuard>,
+}
+
+impl<F, B, E> Future for ResponseFuture<F>
+where
+    F: Future<Output = Result<Response<B>, E>>,
+{
+    type Output = Result<Response<ResponseBody<B>>, E>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+        let response = ready!(this.inner.poll(cx))?;
+        let guard = this.guard.take().unwrap();
+        let response = response.map(move |body| ResponseBody { inner: body, guard });
```

Comment:
The way I'm reading it is that while the "in flight" aspect of a request only starts when it hits this layer -- which is fine, and obviously, you can't know it any earlier -- we actually extend the notion of how long a request is in-flight to the point where the response itself is dropped/consumed.

Personally, this feels wrong to me.  If this was middleware on a client calling a remote service, is the "request" really in flight still if we've received a response but haven't done something with it yet?

Reply:
Good question. I would say yes, if the connecting still open and the body is being read then the request is still in flight. If we've only received the headers but no body then I suppose the request isn't complete. That also applies to streams which might run for a long time.

Reply:
Chatting in Discord, we discussed that based on how a body will typically be used, this approach should be representative of capturing the full lifecycle of a request as it pertains to sending _and_ fully receiving the response, but no further than that point.